File: app/src/main/java/org/freelesson/pagingsample/data/GithubRepository.java
package org.freelesson.pagingsample.data;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import org.freelesson.pagingsample.api.GithubService;
import org.freelesson.pagingsample.db.GithubLocalCache;
import org.freelesson.pagingsample.model.Repo;
import org.freelesson.pagingsample.model.RepoSearchResult;

import java.util.List;

public class GithubRepository {



    private static final int DATABASE_PAGE_SIZE = 20;

    private GithubService service;
    private GithubLocalCache cache;

    public GithubRepository(GithubService service, GithubLocalCache cache) {
        this.service = service;
        this.cache = cache;
    }




    public RepoSearchResult search(String query) {
        Log.d("GithubRepository","New query: "+query);
        DataSource.Factory dataSourceFactory = cache.reposByName(query);
        RepoBoundaryCallback boundaryCallback = new RepoBoundaryCallback(query,service,cache);
        MutableLiveData<String> networkErrors = boundaryCallback.networkErrors;

        LiveData<PagedList<Repo>> data = new LivePagedListBuilder(dataSourceFactory,DATABASE_PAGE_SIZE).setBoundaryCallback(boundaryCallback).build();
        return new RepoSearchResult(data, networkErrors);
    }

}

interface RepositoryCallback<T> {
    void onSuccess(List<Repo> repos);
    void onError(String errorMessage);
}

File: app/src/main/kotlin/org/freelesson/pagingsample/data/GithubRepository.kt
package org.freelesson.pagingsample.data

import androidx.paging.ExperimentalPagingApi
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import kotlinx.coroutines.flow.Flow
import org.freelesson.pagingsample.api.GithubService
import org.freelesson.pagingsample.db.RepoDatabase
import org.freelesson.pagingsample.model.Repo

class GithubRepository(
    private val service: GithubService,
    private val database: RepoDatabase
) {
    @OptIn(ExperimentalPagingApi::class)
    fun getSearchResultStream(query: String): Flow<PagingData<Repo>> {
        val dbQuery = "%${query.replace(' ', '%')}%"
        val pagingSourceFactory = { database.repoDao().reposByName(dbQuery) }
        return Pager(
            config = PagingConfig(
                pageSize = DATABASE_PAGE_SIZE,
                enablePlaceholders = false
            ),
            remoteMediator = GithubRemoteMediator(
                query,
                service,
                database
            ),
            pagingSourceFactory = pagingSourceFactory
        ).flow
    }
}

private const val DATABASE_PAGE_SIZE = 20

internal interface RepositoryCallback<T> {
    fun onSuccess(repos: List<Repo?>?)
    fun onError(errorMessage: String)
}